syntax = "proto3";
package protocol;

import "game_enum.proto";
import "game_struct.proto";

message GameClientEnter {
  uint32 character_id = 1;
  int32 auth_key = 2;
}

message GameServerEnter {
  bool success = 1;
  optional PlayerInfo player_info = 2;
  optional int32 map_id = 3;
}

message GameServerAddPlayer { 
  repeated OtherPlayerInfo player_infos = 1; 
}

message GameServerAddMonster { 
  repeated MobInfo mob_infos = 1;
}

message GameServerRemoveObject { 
  int64 object_id = 1; 
}

message GameServerChangeMap { 
  int32 map_id = 1; 
}

message GameClientPlayerMove {
  float x = 1;
  float y = 2;
  bool flip = 3;
  PlayerAnimation animation = 4;
}

message GameServerPlayerMove {
  int64 object_id = 1;
  float x = 2;
  float y = 3;
  bool flip = 4;
  PlayerAnimation animation = 5;
}

message GameClientChangeKeySetting { 
  KeySetting key_setting = 1; 
}

message GameServerMobMove {
  int64 object_id = 1;
  bool flip = 2;
  MobActionType state = 3;
  float x = 4;
  float y = 5;
  optional uint64 target_id = 6;
  optional float target_x = 7;
  optional float target_y = 8;
}

message GameServerMobAttack {
  int64 target_id = 1;
  int64 mob_id = 2;
}

message GameServerPlayerDamage {
  int64 target_id = 1;
  int32 damage = 2;
}

message GameServerMobDamage {
  int64 target_id = 1;
  int32 damage = 2;
  bool is_critical = 3;
}

message GameClientChat {
  string message = 1;
}

message GameServerChat {
  string message = 1;
  ChatType type = 2;
  optional int64 sender = 3;
}

message GameServerPlayerDead {
  int64 object_id = 1;
}

message GameClientRevive {
}

message GameServerRevive {
  int64 object_id = 1;
}

message GameServerTeleportPlayer {
  int64 object_id = 1;
  float x = 2;
  float y = 3;
}

message GameServerUpdatePlayerStat {
  optional int32 level = 1;
  optional int32 exp = 2;
  optional int32 hp = 3;
  optional int32 mp = 4;
  optional int32 max_hp = 5;
  optional int32 max_mp = 6;
  optional int32 str = 7;
  optional int32 dex = 8;
  optional int32 luk = 9;
  optional int32 int = 10;
  optional int32 ap = 11;
  optional int32 sp = 12;
}